File: imagegenauto/scripts/image_gen_auto.py
"""Completion-first wrapper for the bundled image generation CLI."""

import argparse
import subprocess
import sys
import time
from pathlib import Path


MODELS = ("gpt-image-2", "gpt-image-1.5", "gpt-image-1")
TRANSIENT_MARKERS = ("503", "server error", "temporar", "timeout", "connection", "unavailable")


def value_after(arguments: list[str], flag: str) -> str | None:
    for index, argument in enumerate(arguments[:-1]):
        if argument == flag:
            return arguments[index + 1]
    return None


def remove_model(arguments: list[str]) -> list[str]:
    cleaned: list[str] = []
    skip_next = False
    for argument in arguments:
        if skip_next:
            skip_next = False
        elif argument == "--model":
            skip_next = True
        else:
            cleaned.append(argument)
    return cleaned


def main() -> int:
    parser = argparse.ArgumentParser(description="Retry and fall back across GPT Image models")
    parser.add_argument("command", choices=("generate", "edit"))
    parser.add_argument("arguments", nargs=argparse.REMAINDER)
    parsed = parser.parse_args()

    original_model = value_after(parsed.arguments, "--model")
    models = (original_model,) if original_model else MODELS
    base_arguments = remove_model(parsed.arguments)
    output = value_after(base_arguments, "--out")
    cli = Path(__file__).with_name("image_gen.py")

    for model_index, model in enumerate(models):
        for attempt in range(1, 4):
            command = [sys.executable, str(cli), parsed.command, "--model", model, *base_arguments]
            result = subprocess.run(command, text=True, capture_output=True)
            is_dry_run = "--dry-run" in base_arguments
            if result.returncode == 0 and (
                is_dry_run or not output or Path(output).is_file() and Path(output).stat().st_size > 0
            ):
                sys.stdout.write(result.stdout)
                return 0

            error = (result.stderr + result.stdout).lower()
            if original_model or not any(marker in error for marker in TRANSIENT_MARKERS):
                sys.stderr.write(result.stderr or result.stdout)
                return result.returncode or 1
            if attempt < 3:
                time.sleep(min(8, 2**attempt))
        if model_index < len(models) - 1:
            print(f"{model} unavailable; trying {models[model_index + 1]}", file=sys.stderr)

    print("All configured GPT Image models failed after retries.", file=sys.stderr)
    return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: imagegenauto/scripts/image_gen_auto.mjs
#!/usr/bin/env node
import { spawn } from "node:child_process";
import { access, stat } from "node:fs/promises";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const transient = /408|409|429|500|502|503|504|timeout|temporar|connection|unavailable|server error/i;
const configuredModel = process.env.OPENAI_IMAGE_MODEL;
const models = [...new Set([configuredModel, "gpt-image-2", "gpt-image-1.5", "gpt-image-1"].filter(Boolean))];
const script = resolve(dirname(fileURLToPath(import.meta.url)), "image_gen.mjs");
const args = process.argv.slice(2);
const explicitlySelected = args.includes("--model") ? args[args.indexOf("--model") + 1] : undefined;
const candidates = explicitlySelected
  ? [explicitlySelected]
  : configuredModel?.startsWith("gemini-")
    ? [configuredModel]
    : models;
const outputIndex = args.indexOf("--out");
const output = outputIndex >= 0 ? args[outputIndex + 1] : undefined;

function run(commandArgs) {
  return new Promise(resolveRun => {
    const model = commandArgs[commandArgs.indexOf("--model") + 1];
    process.stderr.write(`[imagegen] ${model}: starting image generation request\n`);
    const child = spawn(process.execPath, [script, ...commandArgs], { stdio: ["inherit", "pipe", "pipe"], env: process.env });
    let stdout = ""; let stderr = "";
    child.stdout.on("data", chunk => { stdout += chunk; });
    child.stderr.on("data", chunk => { stderr += chunk; });
    const waiting = setInterval(() => process.stderr.write(`[imagegen] ${model}: provider is still rendering; waiting for the image response\n`), 15000);
    child.on("error", error => {
      clearInterval(waiting);
      resolveRun({ code: 1, stdout, stderr: `${stderr}\n${error.message}` });
    });
    child.on("close", code => {
      clearInterval(waiting);
      resolveRun({ code: code ?? 1, stdout, stderr });
    });
  });
}

for (const [index, model] of candidates.entries()) {
  for (let attempt = 1; attempt <= 3; attempt += 1) {
    process.stderr.write(`[imagegen] ${model}: attempt ${attempt}/3\n`);
    const forwardedArgs = [];
    for (let i = 1; i < args.length; i += 1) {
      if (args[i] === "--model") {
        i += 1;
        continue;
      }
      forwardedArgs.push(args[i]);
    }
    const commandArgs = [args[0], "--model", model, ...forwardedArgs];
    const result = await run(commandArgs);
    const outputReady = !output || (await access(output).then(() => stat(output)).then(info => info.size > 0).catch(() => false));
    if (result.code === 0 && (args.includes("--dry-run") || outputReady)) { process.stdout.write(result.stdout); process.stderr.write(result.stderr); process.exit(0); }
    const details = `${result.stderr}\n${result.stdout}`;
    if (explicitlySelected || !transient.test(details)) { process.stderr.write(result.stderr || result.stdout); process.exit(result.code || 1); }
    if (attempt < 3) {
      const delay = Math.min(8000, 2 ** attempt * 1000);
      process.stderr.write(`[imagegen] ${model}: transient failure; retrying in ${delay / 1000}s\n`);
      await new Promise(resolveWait => setTimeout(resolveWait, delay));
    }
  }
  if (index < candidates.length - 1) process.stderr.write(`[imagegen] ${model} unavailable; trying ${candidates[index + 1]}\n`);
}
process.stderr.write("All configured GPT Image models failed after retries.\n");
process.exit(1);